### PR TITLE
Minor refactor to the `Plural-Froms` parser

### PR DESCRIPTION
**Pull request by y5nw**
* Rewrite declaration of `reduce_ltr` to check (at compile-time) for mismatching numbers of C++ operator classes and operator pattern strings: https://irc.luanti.org/luanti-dev/2025-09-10#i_6283196
* Replace inner function classes (e.g. `BinaryOperation<T>`) with `std::function` and `std::bind`.
* Fix edge case that can cause client crash.

## How to test

This _should_ be covered by unittests (see `src/unittests/test_translations.cpp`). Please point out whether further testcases should be added.

### CODE REVIEW

File: src/gettext_plural_form.cpp [hunk 1]
```diff
@@ -1,136 +1,77 @@
-// Minetest
+// Luanti
 // SPDX-License-Identifier: LGPL-2.1-or-later
 
+/* This file implements a recursive descent parser for gettext plural forms.
+ * Left recursion (for left-associative operators) is implemented by parse_ltr, which iteratively attempts to reduce
+ * expressions from operations of the same precedence. This should not be confused with reduce_ltr, which recurses
+ * though a list of operators with the same precedence (not the input string!) to search for a match.
```
Comment (SmallJoker) [inline src/gettext_plural_form.cpp]: \* through / throuough ?

File: src/gettext_plural_form.cpp [hunk 2]
```diff
@@ -167,90 +109,101 @@ static ParserResult parse_parenthesized(const size_t nplurals, std::wstring_view
...
+	{
+		return rhs == 0 ? 0 : lhs/rhs;
+	}
+};
+
+template<typename T> struct safe_modulus {
+	T operator()(T lhs, T rhs) const
+	{
+		return rhs == 0 ? 0 : lhs%rhs;
+	}
```
Comment (SmallJoker) [inline src/gettext_plural_form.cpp]: Spaces for readability. I don't think the current notation is well-suited to show up the operator precedence.
```suggestion
		return rhs == 0 ? 0 : (lhs / rhs);
	}
};

template<typename T> struct safe_modulus {
	T operator()(T lhs, T rhs) const
	{
		return rhs == 0 ? 0 : (lhs % rhs);
	}
```

File: src/gettext_plural_form.h [hunk 3]
```diff
@@ -1,33 +1,36 @@
...
-	static GettextPluralForm::Ptr parseHeaderLine(std::wstring_view str);
-protected:
-	GettextPluralForm(size_t nplurals): nplurals(nplurals) {};
+	static Ptr parseHeaderLine(std::wstring_view str) {
+		return Ptr(new GettextPluralForm(str));
+	}
 private:
-	const size_t nplurals;
+	size_t nplurals = 0;
+	Function func = nullptr;
```
Comment (SmallJoker) [inline src/gettext_plural_form.h:44]: While we're at: What does this function do? I can only see that it takes 1 argument and returns a number. Please add a comment.

File: src/gettext_plural_form.cpp (hunk 1 above)
Comment (y5nw) [reply src/gettext_plural_form.cpp]: Corrected in https://github.com/luanti-org/luanti/pull/16489/commits/1175f28cca2e73e3136ded9bfee652cd25f7dcc7 (-> "through")

File: src/gettext_plural_form.h (hunk 3 above)
Comment (y5nw) [reply src/gettext_plural_form.h:44]: https://github.com/luanti-org/luanti/pull/16489/commits/1175f28cca2e73e3136ded9bfee652cd25f7dcc7 / https://github.com/luanti-org/luanti/pull/16489/commits/7a1f3217df6f9f4f3a61964fbf5df24926e06dd1 ?

**Comment by SmallJoker** [comment]
@y5nw I noticed that all valgrind reports have the common line `translation.cpp:638` (using master fcd9c73f2 + PR 7a1f321).
```
loadPoTranslation(basefilename, data);
```
And the only .po file that is loaded is the following: https://codeberg.org/Wuzzy/Repixture/src/branch/master/mods/rp_game_meta/locale/rp_game_meta.de.po
Are you able to reproduce the same memory access issues, using e.g. `valgrind --log-file=valgrind.log --leak-check=full ./luanti` ?

EDIT: Same behaviour observed with "test translations", where valgrind logs warnings in `test_translations.cpp:87`.
